renku_notebooks/api/schemas/server_options.py

```diff
@@ -1,82 +1,159 @@
+            default_url=self.default_url,
+            lfs_auto_fetch=self.lfs_auto_fetch,
+            memory=self.memory / 1000000000,
+            storage=self.storage / 1000000000 if self.storage is not None else None,
+            gigabytes=True,
+        )
+
+    def set_storage(self, storage: int, gigabytes: bool = False):
+        if self.gigabytes and not gigabytes:
+            self.storage = round(storage / 1_000_000_000)
+        elif not self.gigabytes and gigabytes:
+            self.storage = round(storage * 1_000_000_000)
         else:
-            if value != server_options_defaults[field_name]:
-                raise ValidationError(
-                    f"Provided {field_name} value does not match the allowed value of "
-                    f"{server_options_defaults[field_name]}"
-                )
+            self.storage = storage
+
+    def __sub__(self, other: "ServerOptions") -> "ServerOptions":
+        self_storage = 0 if self.storage is None else self.storage
+        other_storage = 0 if other.storage is None else other.storage
+        return ServerOptions(
+            cpu=self.cpu - other.cpu,
+            memoory=self.memory - other.memory,
+            gpu=self.gpu - other.gpu,
+            storage=self_storage - other_storage,
+        )
+
+    def __ge__(self, other: "ServerOptions"):
+        return self.__compare(other, lambda x, y: x >= y)
 
-    return _validate
+    def __gt__(self, other: "ServerOptions"):
+        return self.__compare(other, lambda x, y: x > y)
+
+    def __lt__(self, other: "ServerOptions"):
+        return self.__compare(other, lambda x, y: x < y)
+
+    def __le__(self, other: "ServerOptions"):
+        return self.__compare(other, lambda x, y: x <= y)
+
+    def __eq__(self, other: "ServerOptions"):
+        numeric_value_equal = self.__compare(other, lambda x, y: x == y)
+        return (
+            numeric_value_equal
+            and self.default_url == other.default_url
+            and self.lfs_auto_fetch == other.lfs_auto_fetch
+            and self.gigabytes == other.gigabytes
+            and self.priority_class == other.priority_class
+        )
+
+    @classmethod
+    def from_resource_class(cls, data: Dict[str, Any]) -> "ServerOptions":
+        """Convert a CRC resource class to server options. CRC users GB for storage and memory
+        whereas the notebook service uses bytes so we convert to bytes here."""
+        return cls(
+            cpu=data["cpu"],
+            memory=data["memory"] * 1000000000,
+            gpu=data["gpu"],
+            storage=data["default_storage"] * 1000000000,
+        )
+
+    @classmethod
+    def from_request(cls, data: Dict[str, Any]) -> "ServerOptions":
+        """Convert a server options request dictionary to the model."""
+        return ServerOptions(
+            cpu=data["cpu_request"],
+            gpu=data["gpu_request"],
+            memory=data["mem_request"],
+            default_url=data["defaultUrl"],
```

Comment:
We have a weird mix of camel case and snake case in request fields, we should clean this up at some point

Reply:
I will open an issue for this. We should coordinate with the UI.

Reply:
https://github.com/SwissDataScienceCenter/renku-notebooks/issues/1525